pkg/kubelet/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,363 @@
+		machineInfo, err := kletGetter.GetCachedMachineInfo()
+		if err != nil {
+			return nil, nil, err
+		}
+		topo, err := topology.Discover(machineInfo)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+
+		resources := capacityProvider.GetNodeAllocatableReservation()
+		cpuResource, ok := resources[v1.ResourceCPU]
+		if ok {
+			reservedCores := int(cpuResource.Value())
+			glog.Infof("[cpumanager] reserving %v cores due to kube/system reserved", reservedCores)
+			topo.NumReservedCores = reservedCores
+		}
+		policy = NewStaticPolicy(topo)
+	default:
+		glog.Warningf("[cpumanager] Unknown policy (\"%s\"), falling back to \"%s\" policy (\"%s\")", cpuPolicyName, PolicyNone)
+		policy = NewNonePolicy()
+	}
+
+	manager := &manager{
+		policy:            policy,
+		state:             state.NewMemoryState(),
+		containerRuntime:  cr,
+		kletGetter:        kletGetter,
+		podStatusProvider: statusProvider,
+		getPodsFunc:       getPodsFunc,
+		killPodFunc:       killPodFunc,
+		recorder:          recorder,
+	}
+	return manager, manager, nil
+}
+
+func (m *manager) Start() {
+	glog.Infof("[cpumanger] starting with %s policy", m.policy.Name())
+	m.policy.Start(m.state)
+	if m.policy.Name() == string(PolicyNone) {
+		return
+	}
+	go wait.Until(func() { m.reconcileState() }, time.Second, wait.NeverStop)
+}
+
+func (m *manager) RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	err := m.policy.RegisterContainer(m.state, p, c, containerID)
+	if err != nil {
+		glog.Errorf("[cpumanager] RegisterContainer error: %v", err)
+		return err
+	}
+	cpuset := m.state.GetCPUSetOrDefault(containerID)
+	err = m.containerRuntime.UpdateContainerResources(
+		containerID,
+		&runtimeapi.LinuxContainerResources{
+			CpusetCpus: cpuset.String(),
+		})
+	if err != nil {
+		glog.Errorf("[cpumanager] RegisterContainer error: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (m *manager) UnregisterContainer(containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	err := m.policy.UnregisterContainer(m.state, containerID)
+	if err != nil {
+		glog.Errorf("[cpumanager] UnregisterContainer error: %v", err)
+		return err
+	}
+	return nil
+}
+
+// CPU Manager state is read-only from outside this package.
+func (m *manager) State() state.Reader {
+	return m.state
+}
+
+type reconciledContainer struct {
+	podName       string
+	containerName string
+	containerID   string
+}
+
+func (m *manager) reconcileState() (success []reconciledContainer, failure []reconciledContainer) {
+	m.Lock()
+	defer m.Unlock()
+
+	success = []reconciledContainer{}
+	failure = []reconciledContainer{}
+
+	if m.IsUnderCPUPressure() {
+		// Make sure all no-CPU pods are evicted.
+		// This really should only do something on the first reconcilation
+		// pass after transitioning to UnderPressure conditions, if all
+		// no-CPU pods are evicted successfully.
+		m.evictNoCPUPods()
```

Comment:
Yes, for now. If different requirements come up for later policies we can revisit it. The bigger picture is that it needs to be part of some reconciliation loop because eviction can timeout. So, this avoids the static policy having to run a separate reconcile to manage evictions. One way around this in the future could be to add a `toEvict` list to the state, and use that here.